src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -1499,21 +1499,36 @@ newtype NeutralExpr = NeutralExpr (BackendSyntax NeutralExpr)
 
 derive instance Newtype NeutralExpr _
 
-optimize :: Ctx -> Env -> Qualified Ident -> Int -> BackendExpr -> BackendExpr
-optimize ctx env (Qualified mn (Ident id)) initN = go initN
+optimize :: Boolean -> Ctx -> Env -> Qualified Ident -> Int -> BackendExpr -> Tuple (Maybe (NonEmptyArray BackendExpr)) BackendExpr
+optimize traceSteps ctx env (Qualified mn (Ident id)) initN originalExpr =
+  if traceSteps then goWithSteps (pure originalExpr) initN originalExpr
+  else goNoSteps initN originalExpr
   where
-  go n expr1
+  goNoSteps n expr1 = do
+    let Tuple rewrite expr2 = optimizationStep n expr1
+    if rewrite then
+      goNoSteps (n - 1) expr2
+    else
+      Tuple Nothing expr2
+
+  goWithSteps steps n expr1 = do
+    let Tuple rewrite expr2 = optimizationStep n expr1
+    let updatedSteps = NonEmptyArray.snoc steps expr2
+    if rewrite then
+      goWithSteps updatedSteps (n - 1) expr2
+    else
+      Tuple (Just updatedSteps) expr2
+
+  optimizationStep :: Int -> BackendExpr -> Tuple Boolean BackendExpr
+  optimizationStep n expr1
```

Comment:
I decided to merge these two functions into one so as not to lead to out-of-sync issues with updating one function and forgetting to update the other. 

This function could also be written like below but I thought the extra allocation of an array in the case of tracing being disabled was undesirable:
```purs
optimize ... = go initN <<< pure
  where
  go :: NonEmptyArray BackendExpr -> BackendExpr
  go n acc = do
        let Tuple rewrite expr2 = optimizationStep n $ NEA.last acc
        if rewrite then
            go (n - 1) $ if traceSteps then NEA.snoc acc expr2 else pure expr2
        else
            acc
```